File: apps/embedding-worker/src/dead_letter.py
"""Persistent dead-letter storage for the embedding worker."""
import json
import time
import psycopg2
from src.config import DATABASE_URL


def _log(level: str, event: str, **kwargs):
    entry = {
        "ts": time.strftime("%Y-%m-%dT%H:%M:%S%z"),
        "level": level,
        "service": "embedding-worker",
        "component": "dead_letter",
        "event": event,
    }
    entry.update(kwargs)
    print(json.dumps(entry), flush=True)


def persist_dead_letter(
    job_type: str,
    queue_name: str,
    payload: dict,
    error_message: str,
    stack_trace: str,
    attempt_count: int,
) -> None:
    """Write a dead-letter entry to the Postgres dead_letter_jobs table."""
    try:
        conn = psycopg2.connect(DATABASE_URL, connect_timeout=5)
        try:
            with conn.cursor() as cur:
                cur.execute(
                    """INSERT INTO dead_letter_jobs
                       (job_type, queue_name, payload, error_message, stack_trace, attempt_count, created_at)
                       VALUES (%s, %s, %s, %s, %s, %s, now())""",
                    (
                        job_type,
                        queue_name,
                        json.dumps(payload),
                        error_message,
                        stack_trace,
                        attempt_count,
                    ),
                )
            conn.commit()
        finally:
            conn.close()
    except Exception as e:
        _log("error", "persist_dead_letter_failed", error=str(e), job_type=job_type)

File: apps/truth-worker/src/queue/deadLetter.ts
import type { DeadLetterEntry } from "@hybrid-memory/observability";
import { createLogger } from "@hybrid-memory/observability";
import { getPool } from "../db";

const log = createLogger("truth-worker", "deadLetter");

export async function persistDeadLetter(entry: DeadLetterEntry): Promise<void> {
  try {
    await getPool().query(
      `INSERT INTO dead_letter_jobs (job_type, queue_name, payload, error_message, stack_trace, attempt_count, created_at)
       VALUES ($1, $2, $3, $4, $5, $6, $7)`,
      [
        entry.job_type,
        entry.queue_name,
        JSON.stringify(entry.payload),
        entry.error_message,
        entry.stack_trace,
        entry.attempt_count,
        entry.created_at,
      ]
    );
  } catch (err) {
    log.error("persist_dead_letter_failed", {
      error: err instanceof Error ? err.message : String(err),
      job_type: entry.job_type,
    });
  }
}
